---
title: "Bioinfo Projekt"
author: "Ilya"
date: "5/1/2021"
output: html_document
---

```{r}
load("../2021-topic-01-team-03/dataset/prism_datasets.rda")
load("../2021-topic-01-team-03/dataset/cellline_datasets.rda")
```

# Guidelines

prism: effect of the treatment (columns) on cell growth of cell lines (rows).For example, BRD-A00077618-236-07-6::0.0024414::HTS002 indicates treatment with the drug BRD-A00077618-236-07-6, with a dose of 2.4 nM, in assay HTS002. Values represent log fold changes with respect to a control DMSO experiment.

prism.treat:information about the treatment (row) and the drug used

prism.cl: cell lines' info

prism.exp: gene TPM (transcripts per million) values. Indicate over/underexpression. Rows = cell lines, columns = genes

prism.cnv: gene copy number (CN) values. In absolute terms, CN = 2, since there are two alleles per genes. In cancer, genes might be amplified CN > 2 or deleted CN < 2

prism.snv: observed mutations in a sample. The isDeleterious flag specifies if mutation has a functional effect or not. First column is genes. There is a column
(DepMap_ID) with corresponding cell lines

prism.achilles: gene KO scores, a measure of how essential/important is a particular gene for the cell survival

IDEA: use same indications as in PRISM paper to start filtering for relevant genes, cell lines? 

# Cleaning

## Selecting brain cancer specific cell lines in all data sets

```{r}
brain_cancer_cl=subset.data.frame(prism.cl, disease == 	"Brain Cancer") #identify the brain cancer cell lines' names

names=brain_cancer_cl[,1] #saving the names as names

#select wanted cell lines from all datasets
brain_cancer=prism[names,]
brain_cancer_achilles=prism.achilles[names,]
brain_cancer_cnv=prism.cnv[names,]
brain_cancer_exp=prism.exp[names,]

```

##Selecting relevant info from prism.snv

```{r}
for (i in names){
 result<-which(prism.snv$DepMap_ID == 	
i)
}
result #only delivers results for one of 34 cell lines. without saving it as result it works on all cell lines

sapply(names, function(i){
  snv<-which(prism.snv$DepMap_ID == 	
i)
  View(prism.snv[snv,])
}) #works but kinda weird







```

## NA removal from brain_cancer_achilles

```{r}
brain_cancer_achilles=na.omit(brain_cancer_achilles) #remove the entire cell lines from everywhere without KO scores
```

## NA for mean substituition in brain_cancer

```{r}
for(i in 1:ncol(brain_cancer)){
  brain_cancer[is.na(brain_cancer[,i]), i] <- mean(brain_cancer[,i], na.rm = TRUE)
}
brain_cancer=brain_cancer[ , colSums(is.na(brain_cancer)) == 0]
```

## Brain_cancer distribution visualising

```{r}
max=apply(brain_cancer,2,max)
min=apply(brain_cancer,2,min)
mean=apply(brain_cancer,2,mean)
median=apply(brain_cancer,2,median)
#add quantiles to the dataframe?

drug_summary=data.frame(max, min, mean, median)
sum(is.na(drug_summary$mean))

hist(drug_summary$mean,main= "Distribution of means of each drug",xlab= "means of drugs", freq= F, breaks = 100)
abline(v = quantile(drug_summary$mean, c(0.25,0.5,0.75)), col = "blue", lwd = 2)
#what would be the type of this distribution? Poisson? Negative binominal?

median(drug_summary$mean) # 0.3 criteria from PRISM paper makes sense since it covers more than a half of the drugs
mean(drug_summary$mean) 

qqnorm(drug_summary$mean) #data not normally distributed
abline(a = 0, b = 0.33) #however one part of it seems to match the normal distribution, starting from -1 sample quantile

```

# Question 1: How can we distinguish the most effective drugs?

## Deviding drugs in doses groups

```{r}
dd1=grep("::0.00061034::", colnames(brain_cancer))
dd2=grep("::0.0024414::", colnames(brain_cancer))
dd3=grep("::0.00976562::", colnames(brain_cancer))
dd4=grep("::0.0390625::", colnames(brain_cancer))
dd5=grep("::0.15625::", colnames(brain_cancer))
dd6=grep("::0.625::", colnames(brain_cancer))
dd7=grep("::10::", colnames(brain_cancer))
dd8=grep("::2.5::", colnames(brain_cancer))
#positions of the respective doses in the dataframe 
doses=c(dd1,dd2,dd3,dd4,dd5,dd6,dd7,dd8)


d1=brain_cancer[,dd1]
d2=brain_cancer[,dd2]
d3=brain_cancer[,dd3]
d4=brain_cancer[,dd4]
d5=brain_cancer[,dd5]
d6=brain_cancer[,dd6]
d7=brain_cancer[,dd7]
d8=brain_cancer[,dd8]

#these are the standard doses summed up into seperate dataframes
sum(length(dd1), length(dd2), length(dd3), length(dd4),length(dd5), length(dd6),length(dd7),length(dd8)) #ups something missing

deviation=brain_cancer[,-doses] #lets see what we missed-> these are the doses that are not exact and have some kind of a deviation

questionable_drugs=prism.treat[colnames(deviation),]#we dont know if we can use these drugs, may be their doses deviate too far from the standard
View(questionable_drugs)
```

## Working on drugs that do not fit the standard doses

```{r alternative 1}
#rest of the doeses are all different. What do we do with them?

#lets keep from "deviation" only the ones that have a deviation less than "10%" from original

#to do: make a distribution test for a non normal distribution to justify the 10%
#filtering the doses that deviate less than 10%

keep_d1=which (0.00061034-0.1*0.00061034 <=questionable_drugs[,2] & 0.00061034+0.1*0.00061034 >=questionable_drugs[,2]) #apply the 10% gap and find which genes are that qualify for it
d1_extra=brain_cancer[,rownames(questionable_drugs[keep_d1,])] #create a dataframe with these genes 
d1=cbind(d1,d1_extra) #merge the obtained dataframe with the standard dose one

keep_d2=which (0.0024414-0.1*0.0024414 <= questionable_drugs[,2] & 0.0024414+0.1*0.0024414 >=questionable_drugs[,2])
d2_extra=brain_cancer[,rownames(questionable_drugs[keep_d2,])]
d2=cbind(d2,d2_extra)

keep_d3=which (0.00976562-0.1*0.00976562 <= questionable_drugs[,2] & 0.00976562+0.1*0.00976562 >=questionable_drugs[,2])
d3_extra=brain_cancer[,rownames(questionable_drugs[keep_d3,])]
d3=cbind(d3,d3_extra)

keep_d4=which (0.0390625-0.1*0.0390625 < questionable_drugs[,2] & 0.0390625+0.1*0.0390625 >questionable_drugs[,2])
d4_extra=brain_cancer[,rownames(questionable_drugs[keep_d4,])]
d4=cbind(d4,d4_extra)

keep_d5=which (0.15625-0.1*0.15625 <= questionable_drugs[,2] & 0.15625+0.1*0.15625 >=questionable_drugs[,2])
d5_extra=brain_cancer[,rownames(questionable_drugs[keep_d5,])]
d5=cbind(d5,d5_extra)

keep_d6=which (0.625-0.1*0.625 <= questionable_drugs[,2] & 0.625+0.1*0.625 >=questionable_drugs[,2])
d6_extra=brain_cancer[,rownames(questionable_drugs[keep_d6,])]
d6=cbind(d6,d6_extra)

keep_d7=which (10-0.1*10 <= questionable_drugs[,2] & 10+0.1*10 >=questionable_drugs[,2])
d7_extra=brain_cancer[,rownames(questionable_drugs[keep_d7,])]
d7=cbind(d7,d7_extra)

keep_d8=which (2.5-0.1*2.5 <= questionable_drugs[,2] & 2.5+0.1*2.5 >=questionable_drugs[,2])
d8_extra=brain_cancer[,rownames(questionable_drugs[keep_d8,])]
d8=cbind(d8,d8_extra)


#look at the drugs we did not keep
keep=c(keep_d1, keep_d2, keep_d3, keep_d4,keep_d5,keep_d6,keep_d7,keep_d8)
doses_outliers=deviation[,-keep]

mean(apply(deviation[,-keep],2,mean))#curioisty

```
## Another way to work on drugs that do not fit the standard doses

```{r alternative 2}
#Stefan's suggestion: assigning the outliers to closest standard doses

#Calculation of all differences between questionable drugs and standard doses
std_doses <- c(0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,10,2.5) #vector with standard dosages

#differences between standard values and one exemplary questionable drug

diff <- c()
for (i in std_doses) {
      diff <- abs(i - 0.000610352)
      print(diff)
      #append? How do I can get a vector?
}

# trying something with two nested for loops

#std_doses <- c(0.00061034,0.0024414,0.00976562,0.0390625,0.15625,0.625,10,2.5)
#few_drugs <- c(0.000610352,0.00244141,0.000580005,0.00232002)

#drugs_frame <- data.frame(matrix(nrow = 8, ncol =4))
#for (i in std_doses) {
 # for (j in few_drugs)
     # {
     # drugs_frame[i,j] = abs(i - j)
     #  }
#}
#drugs_frame
```


## Selecting relevant drugs

```{r}
no_keep_drugs=colnames(doses_outliers)#vector with NAMES of drugs we dont want
no_keep_drugs=unique (grep(paste(no_keep_drugs,collapse="|"), colnames(brain_cancer)))#vector of POSITIONS of drugs we dont want in dataframe brain_cancer
brain_cancer=brain_cancer[,-no_keep_drugs]#removing the drugs we dont want
```

## Filter effective drugs in each dose group

```{r}
#Lets say for now that our threshold is 0,3 so we can see how many drugs meet this criteria in each dose group
effective_drugs<-function(x){
  effective_x<-x[,which(apply(x,2,mean)<=0.3)]
  return(effective_x)
}

effective_d1=effective_drugs(d1)
effective_d2=effective_drugs(d2)
effective_d3=effective_drugs(d3)
effective_d4=effective_drugs(d4)
effective_d5=effective_drugs(d5)
effective_d6=effective_drugs(d6)
effective_d7=effective_drugs(d7)
effective_d8=effective_drugs(d8)


all_effective_drugs=c(colnames(effective_d1),colnames(effective_d2),colnames(effective_d3),colnames(effective_d4),colnames(effective_d5),colnames(effective_d6),colnames(effective_d7),colnames(effective_d8))

length(all_effective_drugs)
length(which(apply(brain_cancer,2,mean)<=0.3)) #this is the amount of drugs that we keep in total. We compare these 2 length to make sure that that we do not select the same drug twice. If it was the case the first length would be bigger than the second one
```

##Select the drugs that are effective in multiple doses

```{r}
new_colnames<-function(x){
  a<-gsub(pattern = "::.*", replacement = "", colnames(x))
  return(a)
}
colnames(effective_d1)<-new_colnames(effective_d1)
colnames(effective_d2)<-new_colnames(effective_d2)
colnames(effective_d3)<-new_colnames(effective_d3)
colnames(effective_d4)<-new_colnames(effective_d4)
colnames(effective_d5)<-new_colnames(effective_d5)
colnames(effective_d6)<-new_colnames(effective_d6)
colnames(effective_d7)<-new_colnames(effective_d7)
colnames(effective_d8)<-new_colnames(effective_d8)
## remove everything but the drug information from colnames and make them new colnames of the dataframes

save_new_colnames=function(y){
  colnames(y)<-new_colnames(y)
}
save_new_colnames(effective_d4)
#does not work

  
common_drugs<-function(x,y){
  different<-setdiff(colnames(x), colnames(y))
  
  different<-unique(grep(paste(different,collapse="|"), colnames(x)))
  
  colnames(x[,-different])
}
#function returns the names of the common drugs between 2 dataframes

d1_d2=common_drugs(effective_d1,effective_d2)
d2_d1=common_drugs(effective_d2,effective_d1)
#both are equal, so the function works correctly


d1_d3=common_drugs(effective_d1,effective_d3)
d1_d4=common_drugs(effective_d1,effective_d4)
d1_d5=common_drugs(effective_d1,effective_d5)
d1_d6=common_drugs(effective_d1,effective_d6)
d1_d7=common_drugs(effective_d1,effective_d7)
d1_d8=common_drugs(effective_d1,effective_d8)

#which drugs are present in all 4 8 doses?

present_in_all_doses<-function(a,b,c,d,e,f,g){
 bb<-setdiff(a, b)
 a<-a [! a %in% bb]
 
 cc<-setdiff(a, c)
 a<-a [! a %in% cc]
 
 dd<-setdiff(a, d)
 a<-a [! a %in% dd]
 
 ee<-setdiff(a, e)
 a<-a [! a %in% ee]
 
 ff<-setdiff(a, f)
 a<-a [! a %in% ff]
 
 gg<-setdiff(a, g)
 a<-a [! a %in% gg]
 
 return(a)
}

effective_in_all_doses=present_in_all_doses(d1_d2,d1_d3,d1_d4,d1_d5,d1_d6,d1_d7,d1_d8)
#158 drugs are effective in all doses
```

# Question 2: What are the targets of the effective drugs?

## Apllying the dosis separation of dataset brain_cancer on prism.treat

```{r}
brain_cancer_treat=prism.treat[colnames(brain_cancer),] #selecting only the drugs we previously filtered out from brain_cancer

dt1=brain_cancer_treat[colnames(d1),]
dt2=brain_cancer_treat[colnames(d2),]
dt3=brain_cancer_treat[colnames(d3),]
dt4=brain_cancer_treat[colnames(d4),]
dt5=brain_cancer_treat[colnames(d5),]
dt6=brain_cancer_treat[colnames(d6),]
dt7=brain_cancer_treat[colnames(d7),]
dt8=brain_cancer_treat[colnames(d8),]

#not sure this sorting makes sense now for the brain_cancer_treat, but now we have it in case we need it in the future

summary(brain_cancer_treat)#can be useful to identify trends or come up with some new ideas for filtering, interesting columns can be looked at separetely, for example:
summary(brain_cancer_treat$disease.area)

```

# Question 3: Are there any genetic markers that are specific for brain cancer subtypes?

## Dividing cell lines into subgroups

```{r}
medulloblastoma=brain_cancer_cl$DepMap_ID[grep("Medulloblastoma", brain_cancer_cl$disease_subtype)]#we obtain the names of the cell lines that are classified as medulloblastoma

glioblastoma=brain_cancer_cl$DepMap_ID[grep("Glioblastoma", brain_cancer_cl$disease_subtype)]

glioma=brain_cancer_cl$DepMap_ID[grep("Glioma", brain_cancer_cl$disease_subtype)]

astrocytoma=brain_cancer_cl$DepMap_ID[grep("Astrocytoma", brain_cancer_cl$disease_subtype)]

disease_subtypes=c(medulloblastoma,glioblastoma,glioma,astrocytoma)
length(disease_subtypes)#making sure we didnt forget anything
```

# Question 4: What other factors contribute to drug effectiveness prediction?

# Random: place to try out things

## Clustering

```{r}

brain.km <- kmeans(scale(brain_cancer[, -5]), 6, nstart = 100)
# K-means clusters showing the group of each individuals
brain.km$cluster

```

## PCA for brain_cancer dataset
```{r}
library("factoextra")
dim(brain_cancer)
pca = prcomp(brain_cancer, center = TRUE, scale. = TRUE)

brain.pca <- prcomp(brain_cancer[, -5],  scale = TRUE)
# Coordinates of individuals
coordinates <- as.data.frame(get_pca_ind(brain.pca)$coord)
# Add clusters obtained using the K-means algorithm
coordinates$cluster <- factor(brain.km$cluster)
# Data inspection
head(coordinates)

```

## Analyse the variance of the eigenvalues
```{r}
# Percentage of variance explained by dimensions
eigenvalue <- round(get_eigenvalue(brain.pca), 1)
variance.percent <- eigenvalue$variance.percent
head(eigenvalue)
```


```{r}
library(ggpubr)
ggscatter(
  coordinates, x = "Dim.1", y = "Dim.2",
  color = "cluster", palette = "npg", ellipse = TRUE, ellipse.type = "convex",
  size = 1.5,
  legend = "right",
  ggtheme = theme_bw(),
  xlab = paste0("Dim 1 (", variance.percent[1], "% )" ),
  ylab = paste0("Dim 2 (", variance.percent[2], "% )" )
) +
  stat_mean(aes(color = cluster), size = 3)

```

```{r}

# Compute and plot wss for k = 2 to k = 12
k.max <- 12
data <- coordinates
wss <- sapply(1:k.max,
              function(k){kmeans(data, k, nstart=50,iter.max = 12 )$tot.withinss})
wss
plot(1:k.max, wss,
     type="b", pch = 19, frame = FALSE,
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares")
```




